File: server/src/plugin/sdk/protocol/openai_chat.ts
import type { JsonValue, PluginContext } from "../plugin.ts";
import type { LlmContentPart, LlmRequest, ModelEvent, ProviderOutput } from "../provider.ts";

/** 本协议产生的回放状态种类;与宿主内置 Chat Provider 一致,可互相回放。 */
export const REPLAY_KIND = "openai_chat";

/** 上游返回非 2xx 时抛出,携带完整响应体供调用方分类。 */
export class HttpError extends Error {
  constructor(readonly status: number, readonly body: string) {
    super(`HTTP ${status}: ${body}`);
  }
}

export type OpenAiChatCall = {
  url: string;
  model: string;
  request: LlmRequest;
  headers?: Record<string, string>;
  /** 最后合并进请求体。 */
  extraBody?: Record<string, JsonValue>;
};

function record(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function text(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function count(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

/** 纯文本消息保持字符串形式;混合图片时展开为分块数组。 */
function chatContent(parts: LlmContentPart[]): JsonValue {
  if (parts.every((part) => part.type === "text")) {
    return parts.map((part) => part.type === "text" ? part.text : "").join("");
  }
  return parts.map((part): JsonValue =>
    part.type === "text" ? { type: "text", text: part.text } : {
      type: "image_url",
      image_url: { url: `data:${part.mediaType};base64,${part.dataBase64}` },
    }
  );
}

export function buildChatBody(call: OpenAiChatCall): Record<string, JsonValue> {
  const messages: JsonValue[] = [];
  if (call.request.instructions) {
    messages.push({ role: "system", content: call.request.instructions });
  }
  for (const message of call.request.messages) {
    if (message.role === "assistant") {
      const reasoning = message.replayState?.providerKind === REPLAY_KIND
        ? text(record(message.replayState.value)?.reasoning_content)
        : null;
      // Chat Completions 拒绝空字符串的 assistant content;完全无可见
      // 内容的 assistant 消息不需要发送。
      if (!message.text && message.toolCalls.length === 0 && !reasoning) continue;
      const value: Record<string, JsonValue> = {
        role: "assistant",
        content: message.text ? message.text : null,
      };
      if (reasoning) value.reasoning_content = reasoning;
      if (message.toolCalls.length > 0) {
        value.tool_calls = message.toolCalls.map((toolCall) => ({
          id: toolCall.callId,
          type: "function",
          function: { name: toolCall.name, arguments: JSON.stringify(toolCall.arguments) },
        }));
      }
      messages.push(value);
    } else if (message.role === "tool") {
      messages.push({
        role: "tool",
        content: message.parts.length === 0 ? message.content : chatContent(message.parts),
        tool_call_id: message.callId,
      });
    } else {
      messages.push({ role: message.role, content: chatContent(message.content) });
    }
  }
  const body: Record<string, JsonValue> = {
    model: call.model,
    messages,
    stream: true,
    stream_options: { include_usage: true },
  };
  if (call.request.tools.length > 0) {
    body.tools = call.request.tools.map((tool) => ({
      type: "function",
      function: { name: tool.name, description: tool.description, parameters: tool.parameters },
    }));
  }
  if (call.request.maxOutputTokens !== null) {
    body.max_completion_tokens = call.request.maxOutputTokens;
  }
  if (call.request.reasoning.effort !== null) {
    body.reasoning_effort = call.request.reasoning.effort;
  }
  if (call.request.latency === "fast") body.service_tier = "priority";
  if (call.request.cacheKey !== null) body.prompt_cache_key = call.request.cacheKey;
  return { ...body, ...call.extraBody };
}

type ToolState = {
  callId: string;
  name: string;
  arguments: string;
  emitted: number;
  started: boolean;
};

/** 部分上游会重发完整片段而不是增量;去重后再拼接。 */
function mergeFragment(target: string, fragment: string): string {
  if (target === fragment || target.endsWith(fragment)) return target;
  if (fragment.startsWith(target)) return fragment;
  return target + fragment;
}

function updateTool(
  index: number,
  callId: string | null,
  name: string | null,
  argumentsDelta: string | null,
  tools: Map<number, ToolState>,
): ModelEvent[] {
  let tool = tools.get(index);
  if (!tool) {
    tool = { callId: "", name: "", arguments: "", emitted: 0, started: false };
    tools.set(index, tool);
  }
  if (callId !== null) tool.callId = mergeFragment(tool.callId, callId);
  if (name !== null) tool.name = mergeFragment(tool.name, name);
  if (argumentsDelta !== null) tool.arguments += argumentsDelta;

  const events: ModelEvent[] = [];
  if (!tool.started && tool.callId && tool.name) {
    tool.started = true;
    events.push({ type: "tool-call-start", index, callId: tool.callId, name: tool.name });
  }
  if (tool.started && tool.emitted < tool.arguments.length) {
    events.push({
      type: "tool-call-arguments-delta",
      index,
      delta: tool.arguments.slice(tool.emitted),
    });
    tool.emitted = tool.arguments.length;
  }
  return events;
}

function eventError(value: Record<string, unknown>): string | null {
  const error = value.error;
  if (error === undefined || error === null) return null;
  if (typeof error === "string") return error;
  return text(record(error)?.message) ?? JSON.stringify(error);
}

function usageEvent(value: unknown): ModelEvent {
  const usage = record(value) ?? {};
  return {
    type: "usage",
    usage: {
      inputTokens: count(usage.prompt_tokens),
      outputTokens: count(usage.completion_tokens),
      totalTokens: count(usage.total_tokens),
      cacheReadTokens: count(record(usage.prompt_tokens_details)?.cached_tokens),
      cacheWriteTokens: null,
      reasoningTokens: count(record(usage.completion_tokens_details)?.reasoning_tokens),
    },
  };
}

function mapFinish(reason: string, hasTools: boolean): "stop" | "length" | "tool-use" {
  if (reason === "tool_calls" || reason === "function_call") return "tool-use";
  if (reason === "length") return "length";
  if (reason === "stop" || reason === "content_filter") return "stop";
  return hasTools ? "tool-use" : "stop";
}

async function readBody(lines: AsyncIterable<string>): Promise<string> {
  const collected: string[] = [];
  for await (const line of lines) collected.push(line);
  return collected.join("\n");
}

/**
 * 执行一次 Chat Completions 流式调用,发出与宿主统一事件集一致的标准化事件,
 * 包括文本/思考边界与工具参数增量。非 2xx 响应抛出 `HttpError`,
 * 流内失败抛出 `Error`,由调用方分类额度与授权问题。
 */
export async function streamOpenAiChat(
  call: OpenAiChatCall,
  output: ProviderOutput,
  context: PluginContext,
): Promise<void> {
  const response = await context.network.stream(call.url, {
    method: "POST",
    headers: {
      accept: "text/event-stream",
      "content-type": "application/json",
      ...call.headers,
    },
    body: JSON.stringify(buildChatBody(call)),
  });
  if (response.status < 200 || response.status >= 300) {
    throw new HttpError(response.status, await readBody(response.lines));
  }

  let textOpen = false;
  let thinkingOpen = false;
  let reasoning = "";
  const tools = new Map<number, ToolState>();
  let finalUsage: ModelEvent | null = null;
  let finish: "stop" | "length" | "tool-use" | null = null;
  let sawDoneMarker = false;

  for await (const line of response.lines) {
    if (!line.startsWith("data:")) continue;
    const payload = line.slice(5).trim();
    if (!payload) continue;
    if (payload === "[DONE]") {
      sawDoneMarker = true;
      break;
    }
    let value: Record<string, unknown>;
    try {
      value = record(JSON.parse(payload)) ?? {};
    } catch {
      throw new Error("OpenAI Chat SSE returned invalid JSON");
    }
    const error = eventError(value);
    if (error !== null) throw new Error(`OpenAI Chat error: ${error}`);
    if (value.usage !== undefined && value.usage !== null) {
      finalUsage = usageEvent(value.usage);
    }
    const choice = Array.isArray(value.choices) ? record(value.choices[0]) : null;
    if (!choice) continue;
    const delta = record(choice.delta) ?? {};
    const reasoningDelta = text(delta.reasoning_content) ?? text(delta.reasoning);
    if (reasoningDelta) {
      if (!thinkingOpen) {
        thinkingOpen = true;
        output.emit({ type: "thinking-start" });
      }
      reasoning += reasoningDelta;
      output.emit({ type: "thinking-delta", text: reasoningDelta });
    }
    const content = text(delta.content);
    if (content) {
      if (thinkingOpen) {
        thinkingOpen = false;
        output.emit({ type: "thinking-end" });
      }
      if (!textOpen) {
        textOpen = true;
        output.emit({ type: "text-start" });
      }
      output.emit({ type: "text-delta", text: content });
    }
    if (Array.isArray(delta.tool_calls)) {
      for (const [position, rawTool] of delta.tool_calls.entries()) {
        const toolDelta = record(rawTool);
        if (!toolDelta) continue;
        const index = count(toolDelta.index) ?? position;
        const fn = record(toolDelta.function) ?? {};
        for (
          const event of updateTool(
            index,
            text(toolDelta.id),
            text(fn.name),
            text(fn.arguments),
            tools,
          )
        ) {
          output.emit(event);
        }
      }
    }
    const finishReason = text(choice.finish_reason);
    if (finishReason !== null) finish = mapFinish(finishReason, tools.size > 0);
  }

  if (thinkingOpen) output.emit({ type: "thinking-end" });
  if (textOpen) output.emit({ type: "text-end" });
  for (const [index, tool] of tools) {
    if (!tool.started) {
      if (!tool.name) throw new Error("OpenAI Chat tool call is missing name");
      if (!tool.callId) tool.callId = `call-${index}`;
      tool.started = true;
      output.emit({ type: "tool-call-start", index, callId: tool.callId, name: tool.name });
      if (tool.arguments) {
        tool.emitted = tool.arguments.length;
        output.emit({ type: "tool-call-arguments-delta", index, delta: tool.arguments });
      }
    }
    output.emit({ type: "tool-call-end", index });
  }
  if (finalUsage !== null) output.emit(finalUsage);
  if (reasoning) {
    output.emit({
      type: "replay-state",
      providerKind: REPLAY_KIND,
      value: { reasoning_content: reasoning },
    });
  }
  const reason = finish ??
    (sawDoneMarker ? (tools.size > 0 ? "tool-use" : "stop") : null);
  if (reason === null) {
    throw new Error("OpenAI Chat stream ended without finish_reason");
  }
  output.emit({ type: "done", reason });
}

File: server/src/provider/openai_chat.rs
//! Implements the OpenAI Chat Completions provider adapter.
use std::collections::BTreeMap;

use async_stream::try_stream;
use base64::{engine::general_purpose::STANDARD, Engine};
use eventsource_stream::Eventsource;
use futures_util::StreamExt;
use serde_json::{json, Map, Value};

use crate::{
    config::ProviderConfig,
    model::{
        ContentPart, ModelInvocation, ModelLatency, ProjectedContent, ProjectedMessage, Role,
        ToolCallContent, Usage,
    },
    Error, Result,
};

use super::{
    apply_body_allowlist, apply_openai_prompt_cache_key,
    attempt::{send_once, Attempt},
    map_sse_error, merge_extra_params, provider_event_error,
    recorder::recorded_headers,
    CallRecorder, FinishReason, ModelEvent, Provider, ProviderStream,
};

#[derive(Default)]
struct ChatToolState {
    call_id: String,
    name: String,
    arguments: String,
    emitted_arguments: usize,
    started: bool,
}

pub struct OpenAiChatProvider {
    client: reqwest::Client,
    config: ProviderConfig,
    recorder: Option<CallRecorder>,
}

impl OpenAiChatProvider {
    pub fn new(client: reqwest::Client, config: ProviderConfig) -> Self {
        Self {
            client,
            config,
            recorder: None,
        }
    }

    pub fn with_recorder(mut self, recorder: Option<CallRecorder>) -> Self {
        self.recorder = recorder;
        self
    }
}

impl Provider for OpenAiChatProvider {
    fn stream(
        &self,
        invocation: ModelInvocation,
        cancellation: tokio_util::sync::CancellationToken,
    ) -> ProviderStream {
        let client = self.client.clone();
        let config = self.config.clone();
        let recorder = self.recorder.clone();
        Box::pin(try_stream! {
            let ModelInvocation { call_id, request, .. } = invocation;
            tracing::debug!(
                model = %request.model.model_id,
                call_id = %call_id,
                history_len = request.history.len(),
                tools_count = request.prompt.tools.len(),
                "OpenAI Chat provider stream started"
            );
            let messages = openai_chat_messages(&request.prompt.instructions, &request.history)?;
            let mut body = json!({
                "model": request.model.model_id,
                "messages": messages,
                "stream": true,
                "stream_options": {"include_usage": true}
            });
            if !request.prompt.tools.is_empty() {
                body["tools"] = json!(request.prompt.tools.iter().map(|tool| json!({"type":"function","function":{
                    "name": tool.name, "description": tool.description, "parameters": tool.parameters
                }})).collect::<Vec<_>>());
            }
            apply_model(&mut body, &request.model, config.max_output_tokens)?;
            merge_extra_params(&mut body, &request.model.extra_params)?;
            apply_openai_prompt_cache_key(&mut body, &request.model.model_id)?;
            apply_body_allowlist(&mut body, config.allowed_body_fields.as_ref())?;
            let request_headers = recorded_headers(&config, &[("content-type", "application/json")]);
            if let Some(recorder) = &recorder {
                recorder.request(request_headers.clone(), &body).await?;
            }
            let attempt = send_once(
                "OpenAI Chat",
                || client.post(&config.request_url)
                    .bearer_auth(&config.api_key).headers(config.custom_headers.clone()).json(&body),
                &cancellation,
                recorder.as_ref(),
            ).await?;
            let Attempt::Response(response) = attempt else { return };
            yield ModelEvent::Start { model_call_id: call_id };
            let chunk_recorder = recorder.clone();
            let chunks = response.bytes_stream()
                .map(|chunk| chunk.map_err(Error::from))
                .then(move |chunk| {
                    let recorder = chunk_recorder.clone();
                    async move {
                        let chunk = chunk?;
                        if let Some(recorder) = recorder { recorder.response_chunk(&chunk).await?; }
                        Ok::<_, Error>(chunk)
                    }
                });
            let source = chunks.eventsource();
            futures_util::pin_mut!(source);
            let mut text_open = false;
            let mut thinking_open = false;
            let mut reasoning = String::new();
            let mut tools = BTreeMap::<usize, ChatToolState>::new();
            let mut final_usage = None;
            let mut finish = None;
            let mut saw_done_marker = false;
            let mut loop_iteration: u64 = 0;
            loop {
                loop_iteration += 1;
                let event = tokio::select! {
                    _ = cancellation.cancelled() => {
                        tracing::debug!(
                            iteration = loop_iteration,
                            saw_done_marker,
                            tool_count = tools.len(),
                            "OpenAI Chat stream cancelled"
                        );
                        return;
                    }
                    event = source.next() => event,
                };
                let Some(event) = event else {
                    tracing::debug!(
                        iteration = loop_iteration,
                        saw_done_marker,
                        "OpenAI Chat SSE stream ended"
                    );
                    break;
                };
                let event = event.map_err(|error| {
                    tracing::debug!(iteration = loop_iteration, error = %error, "OpenAI Chat SSE event failed");
                    map_sse_error("OpenAI Chat", error)
                })?;
                if event.data == "[DONE]" { saw_done_marker = true; break; }
                let value: Value = serde_json::from_str(&event.data)?;
                if let Some(error) = provider_event_error("OpenAI Chat", &value) {
                    Err(error)?;
                }
                if let Some(usage) = value.get("usage").filter(|value| !value.is_null()) {
                    final_usage = Some(openai_usage(usage));
                }
                let Some(choice) = value.get("choices").and_then(Value::as_array).and_then(|values| values.first()) else { continue; };
                let delta = choice.get("delta").unwrap_or(&Value::Null);
                if let Some(reasoning_delta) = delta.get("reasoning_content").or_else(|| delta.get("reasoning")).and_then(Value::as_str).filter(|text| !text.is_empty()) {
                    if !thinking_open { thinking_open = true; yield ModelEvent::ThinkingStart; }
                    reasoning.push_str(reasoning_delta);
                    yield ModelEvent::ThinkingDelta(reasoning_delta.into());
                }
                if let Some(content) = delta.get("content").and_then(Value::as_str).filter(|text| !text.is_empty()) {
                    if thinking_open { thinking_open = false; yield ModelEvent::ThinkingEnd; }
                    if !text_open { text_open = true; yield ModelEvent::TextStart; }
                    yield ModelEvent::TextDelta(content.into());
                }
                if let Some(tool_deltas) = delta.get("tool_calls").and_then(Value::as_array) {
                    for (position, tool) in tool_deltas.iter().enumerate() {
                        let index = tool.get("index").and_then(Value::as_u64).map_or(position, |index| index as usize);
                        let id = tool.get("id").and_then(Value::as_str);
                        let function = tool.get("function").unwrap_or(&Value::Null);
                        let name = function.get("name").and_then(Value::as_str);
                        let arguments = function.get("arguments").and_then(Value::as_str);
                        for event in update_chat_tool(index, id, name, arguments, &mut tools) { yield event; }
                    }
                }
                if let Some(reason) = choice.get("finish_reason").and_then(Value::as_str) {
                    finish = Some(map_finish(reason, !tools.is_empty()));
                }
            }
            if thinking_open { yield ModelEvent::ThinkingEnd; }
            if text_open { yield ModelEvent::TextEnd; }
            for (index, tool) in &mut tools {
                if !tool.started {
                    if tool.name.is_empty() {
                        Err(Error::Provider("OpenAI Chat tool call is missing name".into()))?;
                    }
                    if tool.call_id.is_empty() {
                        tool.call_id = format!("call-{index}");
                    }
                    tool.started = true;
                    yield ModelEvent::ToolCallStart { index: *index, call_id: tool.call_id.clone(), name: tool.name.clone() };
                    if !tool.arguments.is_empty() {
                        tool.emitted_arguments = tool.arguments.len();
                        yield ModelEvent::ToolCallArgumentsDelta { index: *index, delta: tool.arguments.clone() };
                    }
                }
                yield ModelEvent::ToolCallEnd { index: *index };
            }
            if let Some(usage) = final_usage { yield ModelEvent::Usage(usage); }
            if !reasoning.is_empty() {
                yield ModelEvent::ProviderReplayState(crate::model::ProviderReplayState {
                    provider_kind: "openai_chat".into(),
                    value: json!({"reasoning_content": reasoning}),
                });
            }
            let finish = finish.or_else(|| saw_done_marker.then_some(if tools.is_empty() { FinishReason::Stop } else { FinishReason::ToolUse }))
                .ok_or_else(|| Error::Provider("OpenAI Chat stream ended without finish_reason".into()))?;
            yield ModelEvent::Done(finish);
        })
    }
}

fn apply_model(
    body: &mut Value,
    model: &crate::model::ModelSpec,
    route_max_output_tokens: Option<u64>,
) -> Result<()> {
    let object = body
        .as_object_mut()
        .ok_or_else(|| Error::Provider("OpenAI Chat request body is not an object".into()))?;
    if let Some(max) = model.max_output_tokens.or(route_max_output_tokens) {
        object.insert("max_completion_tokens".into(), json!(max));
    }
    if let Some(effort) = &model.reasoning.effort {
        object.insert("reasoning_effort".into(), json!(effort));
    }
    if model.latency == ModelLatency::Fast {
        object.insert("service_tier".into(), json!("priority"));
    }
    Ok(())
}

fn openai_chat_messages(instructions: &str, messages: &[ProjectedMessage]) -> Result<Vec<Value>> {
    let mut output = Vec::with_capacity(messages.len() + usize::from(!instructions.is_empty()));
    if !instructions.is_empty() {
        output.push(json!({"role": "system", "content": instructions}));
    }
    for message in messages {
        let mut value = Map::new();
        value.insert(
            "role".into(),
            Value::String(role_name(&message.role).into()),
        );
        match &message.content {
            ProjectedContent::Parts(parts) => {
                value.insert("content".into(), chat_content(&message.role, parts)?);
            }
            ProjectedContent::Assistant {
                text,
                replay_state,
                calls,
                ..
            } => {
                let replay_reasoning = replay_state
                    .as_ref()
                    .filter(|state| state.provider_kind == "openai_chat")
                    .and_then(|state| state.value.get("reasoning_content"))
                    .and_then(Value::as_str)
                    .filter(|reasoning| !reasoning.is_empty());

                // Chat Completions rejects an empty assistant content string. Tool-call
                // assistant messages use null content, while an assistant with no visible
                // content at all does not need to be sent.
                if text.is_empty() && calls.is_empty() && replay_reasoning.is_none() {
                    continue;
                }
                value.insert(
                    "content".into(),
                    if text.is_empty() {
                        Value::Null
                    } else {
                        Value::String(text.clone())
                    },
                );
                if let Some(reasoning) = replay_reasoning {
                    value.insert("reasoning_content".into(), Value::String(reasoning.into()));
                }
                if !calls.is_empty() {
                    value.insert(
                        "tool_calls".into(),
                        Value::Array(
                            calls
                                .iter()
                                .map(openai_tool_call)
                                .collect::<Result<Vec<_>>>()?,
                        ),
                    );
                }
            }
            ProjectedContent::ToolResult(result) => {
                value.insert(
                    "content".into(),
                    if result.provider_parts.is_empty() {
                        Value::String(result.content.clone())
                    } else {
                        chat_content(&Role::User, &result.provider_parts)?
                    },
                );
                value.insert("tool_call_id".into(), Value::String(result.call_id.clone()));
            }
        }
        output.push(Value::Object(value));
    }
    Ok(output)
}

fn chat_content(_role: &Role, parts: &[ContentPart]) -> Result<Value> {
    let mut text = String::new();
    let mut only_text = true;
    for part in parts {
        match part {
            ContentPart::Text { text: part } => text.push_str(part),
            ContentPart::Image { .. } => {
                only_text = false;
                break;
            }
        }
    }
    if only_text {
        return Ok(Value::String(text));
    }
    Ok(Value::Array(
        parts
            .iter()
            .map(|part| match part {
                ContentPart::Text { text } => Ok(json!({"type":"text", "text":text})),
                ContentPart::Image { mime_type, data } => Ok(json!({
                    "type":"image_url",
                    "image_url":{"url":format!(
                        "data:{mime_type};base64,{}",
                        STANDARD.encode(data)
                    )},
                })),
            })
            .collect::<Result<Vec<_>>>()?,
    ))
}

fn openai_tool_call(call: &ToolCallContent) -> Result<Value> {
    Ok(json!({
        "id": call.call_id,
        "type": "function",
        "function": {
            "name": call.name,
            "arguments": serde_json::to_string(&call.arguments)?,
        }
    }))
}

fn role_name(role: &Role) -> &'static str {
    match role {
        Role::System => "system",
        Role::User => "user",
        Role::Assistant => "assistant",
        Role::Tool => "tool",
    }
}

fn map_finish(value: &str, has_tools: bool) -> FinishReason {
    match value {
        "tool_calls" | "function_call" => FinishReason::ToolUse,
        "length" => FinishReason::Length,
        "content_filter" => FinishReason::Stop,
        // Observed tool calls outrank ordinary or unknown stop labels because
        // OpenAI-compatible servers commonly emit tools with `stop`.
        _ if has_tools => FinishReason::ToolUse,
        _ => FinishReason::Stop,
    }
}

fn update_chat_tool(
    index: usize,
    call_id: Option<&str>,
    name: Option<&str>,
    arguments: Option<&str>,
    tools: &mut BTreeMap<usize, ChatToolState>,
) -> Vec<ModelEvent> {
    let tool = tools.entry(index).or_default();
    if let Some(call_id) = call_id {
        merge_chat_fragment(&mut tool.call_id, call_id);
    }
    if let Some(name) = name {
        merge_chat_fragment(&mut tool.name, name);
    }
    if let Some(arguments) = arguments {
        tool.arguments.push_str(arguments);
    }

    let mut events = Vec::new();
    if !tool.started && !tool.call_id.is_empty() && !tool.name.is_empty() {
        tool.started = true;
        events.push(ModelEvent::ToolCallStart {
            index,
            call_id: tool.call_id.clone(),
            name: tool.name.clone(),
        });
    }
    if tool.started && tool.emitted_arguments < tool.arguments.len() {
        let delta = tool.arguments[tool.emitted_arguments..].to_string();
        tool.emitted_arguments = tool.arguments.len();
        events.push(ModelEvent::ToolCallArgumentsDelta { index, delta });
    }
    events
}

fn merge_chat_fragment(target: &mut String, fragment: &str) {
    if target == fragment || target.ends_with(fragment) {
        return;
    }
    if fragment.starts_with(target.as_str()) {
        *target = fragment.into();
    } else {
        target.push_str(fragment);
    }
}

pub(crate) fn openai_usage(value: &Value) -> Usage {
    let input_tokens = value.get("prompt_tokens").and_then(Value::as_u64);
    Usage {
        input_tokens,
        context_input_tokens: input_tokens,
        output_tokens: value.get("completion_tokens").and_then(Value::as_u64),
        total_tokens: value.get("total_tokens").and_then(Value::as_u64),
        cache_read_tokens: value
            .pointer("/prompt_tokens_details/cached_tokens")
            .and_then(Value::as_u64),
        cache_write_tokens: None,
        reasoning_tokens: value
            .pointer("/completion_tokens_details/reasoning_tokens")
            .and_then(Value::as_u64),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fast_latency_uses_openai_priority_service_tier() {
        let mut body = json!({});
        let mut model = crate::model::ModelSpec::new("test-model");
        model.latency = ModelLatency::Fast;

        apply_model(&mut body, &model, None).unwrap();

        assert_eq!(body["service_tier"], "priority");
    }

    #[test]
    fn observed_tool_calls_outrank_ordinary_stop_reasons() {
        assert_eq!(map_finish("stop", true), FinishReason::ToolUse);
        assert_eq!(map_finish("", true), FinishReason::ToolUse);
    }

    #[test]
    fn content_filter_never_executes_observed_tool_calls() {
        assert_eq!(map_finish("content_filter", true), FinishReason::Stop);
        assert_eq!(map_finish("content_filter", false), FinishReason::Stop);
    }

    #[test]
    fn finish_reason_mapping_without_tool_calls_is_unchanged() {
        assert_eq!(map_finish("stop", false), FinishReason::Stop);
        assert_eq!(map_finish("content_filter", false), FinishReason::Stop);
        assert_eq!(map_finish("", false), FinishReason::Stop);
        assert_eq!(map_finish("tool_calls", false), FinishReason::ToolUse);
        assert_eq!(map_finish("function_call", false), FinishReason::ToolUse);
    }

    #[test]
    fn a_truncated_response_stays_truncated_even_with_tool_calls() {
        assert_eq!(map_finish("length", true), FinishReason::Length);
        assert_eq!(map_finish("length", false), FinishReason::Length);
    }
}
